framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,481 @@
 
-        GenericAnnotatedTypeFactory<V, ?, ?, ?> factory = analysis.getTypeFactory();
-        ProcessingEnvironment processingEnv = factory.getProcessingEnv();
-        AnnotatedTypeMirror lubAnnotatedType =
-                AnnotatedTypes.leastUpperBound(processingEnv, factory, type, otherType);
+        if (missingHierarchy != null) {
+            String missing = PluginUtil.join(", ", missingHierarchy);
+            ErrorReporter.errorAbort(
+                    "CFAbstractValue: missing annotation in the following hierarchies: " + missing);
+            return false;
+        }
+        return true;
+    }
 
-        return analysis.createAbstractValue(lubAnnotatedType);
+    /**
+     * Returns whether or not the set of annotations can be missing an annotation for any hierarchy
+     * @return whether or not the set of annotations can be missing an annotation
+     */
+    public boolean canBeMissingAnnotations() {
+        return canBeMissingAnnotations(underlyingType);
     }
 
-    private static void copyArrayComponentAnnotations(
-            AnnotatedArrayType source, AnnotatedArrayType dest) {
-        AnnotatedTypeMirror destComp = dest.getComponentType();
-        AnnotatedTypeMirror sourceComp = source.getComponentType();
-        destComp.addAnnotations(sourceComp.getAnnotations());
-        if (sourceComp instanceof AnnotatedArrayType) {
-            assert dest instanceof AnnotatedArrayType;
-            copyArrayComponentAnnotations(
-                    (AnnotatedArrayType) sourceComp, (AnnotatedArrayType) destComp);
+    private boolean canBeMissingAnnotations(TypeMirror typeMirror) {
+        if (typeMirror == null) {
+            return false;
         }
+        if (typeMirror.getKind() == TypeKind.WILDCARD) {
+            return canBeMissingAnnotations(((WildcardType) typeMirror).getExtendsBound());
+        }
+        return typeMirror.getKind() == TypeKind.TYPEVAR;
     }
 
     /**
-     * Returns the annotations on the upper bound of type {@code t}.
+     * Returns a set of annotations.  If {@link #canBeMissingAnnotations()} returns true, then
+     * the set of annotations may not have an annotation for every hierarchy.
+     * @return Returns a set of annotations
      */
-    //TODO_JB: INTERSECTIONS AREN'T TAKEN CARE OF
-    private static Collection<AnnotationMirror> getUpperBound(AnnotatedTypeMirror t) {
-        if (t.getKind() == TypeKind.WILDCARD) {
-            AnnotatedTypeMirror extendsBound = ((AnnotatedWildcardType) t).getExtendsBound();
-            if (extendsBound != null) {
-                return extendsBound.getEffectiveAnnotations();
-            }
-        } else if (t.getKind() == TypeKind.TYPEVAR) {
-            AnnotatedTypeMirror upperBound = ((AnnotatedTypeVariable) t).getUpperBound();
-            if (upperBound != null) {
-                return upperBound.getEffectiveAnnotations();
-            }
+    @Pure // because the annotations field is final
```

Comment:
The `because` comment doesn't make sense. The method only returns the field. Even if the field is not final, this method is pure. Same for `getUnderlyingType` below.